src/systemdhook.c

```diff
@@ -332,57 +332,73 @@ static int move_mount_to_runtmp(const char *rootfs, const char *run_tmp_dir, con
 	return 0;
 }
 
-#define RUN_PREFIX "/run/"
+static int move_mounts(const char *rootfs,
+		       const char *path,
+		       const char **config_mounts,
+		       unsigned config_mounts_len,
+		       char *options
+	) {
 
-/*
- * Move the mounts under /run to temporary directory under /tmp (on the host)
- */
-static int adjust_run_mounts(const char *rootfs,
-		 const char *run_tmp_dir,
-		 const char **config_mounts,
-		 unsigned len)
-{
 	int rc = -1;
-	struct stat secrets_st = {0};
-	struct stat run_st = {0};
-	_cleanup_free_ char *run_secrets_dir = NULL;
-	_cleanup_free_ char *run_dir = NULL;
+	char mount_dir[PATH_MAX];
+	snprintf(mount_dir, PATH_MAX, "%s%s", rootfs, path);
 
-	rc = asprintf(&run_dir, "%s/%s", rootfs, "run");
-	if (rc < 0) {
-		pr_perror("Failed to allocate memory for run_dir");
-		return -1;
-	}
+	/* Create a temporary directory to move the PATH mounts to */
+	char temp_template[] = "/tmp/ocitmp.XXXXXX";
 
-	rc = asprintf(&run_secrets_dir, "%s/%s", rootfs, "run/secrets");
-	if (rc < 0) {
-		pr_perror("Failed to allocate memory for run_secrets_dir");
+	char *tmp_dir = mkdtemp(temp_template);
+	if (tmp_dir == NULL) {
+		pr_perror("Failed to create temporary directory for mounts");
 		return -1;
 	}
 
-	/* Move /run/secrets to temporary directory if it exists and it is a mountpoint */
-	if (stat(run_dir, &run_st) == 0 &&
-	    stat(run_secrets_dir, &secrets_st) == 0 &&
-	    secrets_st.st_rdev != run_st.st_rdev &&
-	    (S_ISDIR(secrets_st.st_mode))) {
-		if (move_mount_to_runtmp(rootfs, run_tmp_dir, "/run/secrets") < 0) {
-			pr_perror("Failed to move %s to %s", run_secrets_dir, run_tmp_dir);
+	/* Create the PATH directory */
+	if (!contains_mount(config_mounts, config_mounts_len, path)) {
+		if (mkdir(mount_dir, 0755) == -1) {
+			if (errno != EEXIST) {
+				pr_perror("Failed to mkdir: %s", mount_dir);
+				return -1;
+			}
+		}
+
+		/* Mount tmpfs at new temp directory */
+		if (mount("tmpfs", tmp_dir, "tmpfs", MS_NODEV|MS_NOSUID, options) == -1) {
+			pr_perror("Failed to mount tmpfs at %s", tmp_dir);
 			return -1;
 		}
-	}
 
-	/* Move other user specified mounts under /run to temporary directory */
-	for (unsigned i = 0; i < len; i++) {
-		/* Match destinations that begin with /run/ */
-		if (!strncmp(RUN_PREFIX, config_mounts[i], strlen(RUN_PREFIX))) {
-			if (move_mount_to_runtmp(rootfs, run_tmp_dir, config_mounts[i]) < 0) {
-				pr_perror("Failed to move %s to %s", config_mounts[i], run_tmp_dir);
+		/* Special case for /run/secrets which will not be in the 
+		   config_mounts */
```

Comment:
We need to retain the check to see if /run/secrets is a mount point before we attempt to move it. It was there in the eariler code.